File: app/src/main/java/com/arielfaridja/ezrahi/data/DataRepoFactory.java
package com.arielfaridja.ezrahi.data;

import android.content.Context;

public class DataRepoFactory {
    static IDataRepo instance = null;

    private DataRepoFactory() {
    }

    public static IDataRepo getInstance(Context context) {
        if (instance == null) {
            instance = new FirebaseDataRepo(context);
        }

        return instance;
    }

    public static IDataRepo getInstance() {
        return instance;
    }
}

File: app/src/main/java/com/arielfaridja/ezrahi/data/DataRepoFactory.kt
package com.arielfaridja.ezrahi.data

import android.content.Context
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.components.SingletonComponent

object DataRepoFactory {
    private var instance: IDataRepo? = null

    fun getInstance(context: Context): IDataRepo {
        if (instance == null) {
            instance = EntryPointAccessors.fromApplication(context, DataRepoEntryPoint::class.java).dataRepo()
        }
        return instance!!
    }

    fun getInstance(): IDataRepo = instance
        ?: throw IllegalStateException("DataRepoFactory not initialized. Call getInstance(context) first.")
}

@EntryPoint
@InstallIn(SingletonComponent::class)
interface DataRepoEntryPoint {
    fun dataRepo(): IDataRepo
}